// curve/lib/rectangle-selection.js
(function() {
  var CompositeDisposable, RectangleSelection,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  CompositeDisposable = require('event-kit').CompositeDisposable;

  module.exports = RectangleSelection = (function() {
    function RectangleSelection(svgDocument, options) {
      var base;
      this.svgDocument = svgDocument;
      this.options = options != null ? options : {};
      this.render = bind(this.render, this);
      if ((base = this.options)["class"] == null) {
        base["class"] = 'object-selection';
      }
    }

    RectangleSelection.prototype.setObject = function(object) {
      if (object === this.object) {
        return;
      }
      this._unbindObject();
      this.object = object;
      this._bindObject(this.object);
      if (this.trackingObject) {
        this.trackingObject.remove();
      }
      this.trackingObject = null;
      if (this.object) {
        this.trackingObject = this.svgDocument.getToolLayer().rect();
        this.trackingObject.node.setAttribute('class', this.options["class"] + ' invisible-to-hit-test');
        this.trackingObject.back();
        this.render();
      }
    };

    RectangleSelection.prototype.render = function() {
      var position, size;
      position = this.object.get('position');
      size = this.object.get('size');
      return this.trackingObject.attr({
        x: Math.round(position.x) - .5,
        y: Math.round(position.y) - .5,
        width: Math.round(size.width) + 1,
        height: Math.round(size.height) + 1,
        transform: this.object.get('transform')
      });
    };

    RectangleSelection.prototype._bindObject = function(object) {
      if (!object) {
        return;
      }
      this.selectedObjectSubscriptions = new CompositeDisposable;
      return this.selectedObjectSubscriptions.add(object.on('change', this.render));
    };

    RectangleSelection.prototype._unbindObject = function() {
      var ref;
      if ((ref = this.selectedObjectSubscriptions) != null) {
        ref.dispose();
      }
      return this.selectedObjectSubscriptions = null;
    };

    return RectangleSelection;

  })();

}).call(this);
