// curve/lib/rectangle.js
(function() {
  var DefaultAttrs, Delegator, Draggable, Emitter, IDS, ObjectAssign, Rectangle, RectangleModel, Utils,
    bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; };

  Emitter = require('event-kit').Emitter;

  ObjectAssign = require('object-assign');

  Delegator = require('delegato');

  Utils = require('./utils');

  Draggable = require('./draggable-mixin');

  RectangleModel = require('./rectangle-model');

  DefaultAttrs = {
    x: 0,
    y: 0,
    width: 10,
    height: 10,
    fill: '#eee',
    stroke: 'none'
  };

  IDS = 0;

  module.exports = Rectangle = (function() {
    Draggable.includeInto(Rectangle);

    Delegator.includeInto(Rectangle);

    Rectangle.delegatesMethods('on', {
      toProperty: 'emitter'
    });

    Rectangle.delegatesMethods('get', 'set', 'getID', 'getType', 'toString', 'translate', {
      toProperty: 'model'
    });

    function Rectangle(svgDocument1, options) {
      this.svgDocument = svgDocument1;
      if (options == null) {
        options = {};
      }
      this.onModelChange = bind(this.onModelChange, this);
      this.emitter = new Emitter;
      this.model = new RectangleModel;
      this._setupSVGObject(options);
      this.model.on('change', this.onModelChange);
      this.svgDocument.registerObject(this);
    }


    /*
    Section: Public Methods
     */

    Rectangle.prototype.remove = function() {
      this.svgEl.remove();
      return this.emitter.emit('remove', {
        object: this
      });
    };

    Rectangle.prototype.updateFromAttributes = function() {
      var fill, height, transform, width, x, y;
      x = this.svgEl.attr('x');
      y = this.svgEl.attr('y');
      width = this.svgEl.attr('width');
      height = this.svgEl.attr('height');
      transform = this.svgEl.attr('transform');
      fill = this.svgEl.attr('fill');
      return this.model.set({
        position: [x, y],
        size: [width, height],
        transform: transform,
        fill: fill
      });
    };

    Rectangle.prototype.render = function(svgEl) {
      var position, size;
      if (svgEl == null) {
        svgEl = this.svgEl;
      }
      position = this.model.get('position');
      size = this.model.get('size');
      return svgEl.attr({
        x: position.x,
        y: position.y,
        width: size.width,
        height: size.height,
        transform: this.model.get('transform') || null,
        fill: this.model.get('fill') || null
      });
    };

    Rectangle.prototype.cloneElement = function(svgDocument) {
      var el;
      if (svgDocument == null) {
        svgDocument = this.svgDocument;
      }
      el = svgDocument.getObjectLayer().rect();
      this.render(el);
      return el;
    };


    /*
    Section: Event Handlers
     */

    Rectangle.prototype.onModelChange = function(args) {
      this.render();
      args.object = this;
      return this.emitter.emit('change', args);
    };


    /*
    Section: Private Methods
     */

    Rectangle.prototype._setupSVGObject = function(options) {
      this.svgEl = options.svgEl;
      if (!this.svgEl) {
        this.svgEl = this.svgDocument.getObjectLayer().rect().attr(ObjectAssign({}, DefaultAttrs, options));
      }
      Utils.setObjectOnNode(this.svgEl.node, this);
      return this.updateFromAttributes();
    };

    return Rectangle;

  })();

}).call(this);
